### PR TITLE
Add PSRAM mode and speed config

### PR DESCRIPTION
**Pull request by kbx81**
# What does this implement/fix?

Adds params `mode` and `speed` to enable easy configuration of the respective PSRAM parameters via `sdkconfig`.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3177

## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

```yaml
psram:
  mode: octal
  speed: 80MHz
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/psram/__init__.py [hunk 1]
```diff
@@ -4,15 +4,35 @@
...
 
+SPIRAM_MODES = {
+    "quad": "CONFIG_SPIRAM_MODE_QUAD",
+    "oct": "CONFIG_SPIRAM_MODE_OCT",
+}
+
+SPIRAM_SPEEDS = {
+    "40m": "CONFIG_SPIRAM_SPEED_40M",
+    "80m": "CONFIG_SPIRAM_SPEED_80M",
+    "120m": "CONFIG_SPIRAM_SPEED_120M",
```
Comment (jesserockz) [inline esphome/components/psram/__init__.py]:
```suggestion
    40e6: "CONFIG_SPIRAM_SPEED_40M",
    80e6": "CONFIG_SPIRAM_SPEED_80M",
    120e6": "CONFIG_SPIRAM_SPEED_120M",
```

File: esphome/components/psram/__init__.py [hunk 2]
```diff
@@ -4,15 +4,35 @@
...
+}
+
 CONFIG_SCHEMA = cv.All(
-    cv.Schema({cv.GenerateID(): cv.declare_id(PsramComponent)}), cv.only_on_esp32
+    cv.Schema(
+        {
+            cv.GenerateID(): cv.declare_id(PsramComponent),
+            cv.Optional(CONF_MODE): cv.enum(SPIRAM_MODES, lower=True),
+            cv.Optional(CONF_SPEED): cv.enum(SPIRAM_SPEEDS, lower=True),
```
Comment (jesserockz) [inline esphome/components/psram/__init__.py]: This combine with the above change might be a little nicer
```suggestion
            cv.Optional(CONF_SPEED): cv.all(cv.frequency, cv.one_of(*SPIRAM_SPEEDS)),
```

File: esphome/components/psram/__init__.py [hunk 3]
```diff
@@ -4,15 +4,35 @@
...
 )
 
 CODEOWNERS = ["@esphome/core"]
 
 psram_ns = cg.esphome_ns.namespace("psram")
 PsramComponent = psram_ns.class_("PsramComponent", cg.Component)
 
+SPIRAM_MODES = {
+    "quad": "CONFIG_SPIRAM_MODE_QUAD",
+    "oct": "CONFIG_SPIRAM_MODE_OCT",
```
Comment (jesserockz) [inline esphome/components/psram/__init__.py]: Maybe full name for it here?
```suggestion
    "octal": "CONFIG_SPIRAM_MODE_OCT",
```

Comment (kbx81) [reply esphome/components/psram/__init__.py]: SGTM!

**Review by jesserockz — APPROVED** [review]
Working well on s3 box with octal and 80mhz
```yaml
psram:
  mode: octal
  speed: 80MHz
```

**Comment by jesserockz** [comment]
We probably should add a docs page about psram now come to think of it...

**Comment by martijnbuts** [comment]
> We probably should add a docs page about psram now come to think of it...

Would it be a option to also add the sensor for it to the documentation page?
Then people know if the psram in available and how much it is used. 

  - platform: template
    name: Free psram
    lambda: return ESP.getFreePsram();
    icon: "mdi:memory"
    entity_category: diagnostic
    state_class: measurement
    unit_of_measurement: "b"
    update_interval: 60s

**Pull request merged**